File: Array/2948. Make Lexicographically Smallest Array by Swapping Elements.cpp
class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) 
    {
        vector<int> ans = nums;
        sort(begin(nums), end(nums));

        int parent = 0, n = nums.size();
        unordered_map<int, int> valueParent;
        unordered_map<int, deque<int>> store;

        valueParent[nums[0]] = parent;
        store[parent].push_back(nums[0]);
        for(int i = 1; i < n; i++)
        {
            if((nums[i] - nums[i - 1]) > limit) parent += 1;

            valueParent[nums[i]] = parent;
            store[parent].push_back(nums[i]);
        }

        for(auto &val : ans)
        {
            parent = valueParent[val];
            val = store[parent].front();
            store[parent].pop_front();
        }
        return ans;
    }
};

File: Deque/2948. Make Lexicographically Smallest Array by Swapping Elements.py
class Solution:
    def lexicographicallySmallestArray(self, nums: List[int], limit: int) -> List[int]:
        nums_cp = list(nums)
        nums.sort()

        n = len(nums)
        parent_pos = 0

        parent = defaultdict(int)
        store = defaultdict(deque)

        parent[nums[0]] = 0
        store[parent_pos].append(nums[0])
        
        for i in range(1, n):
            if nums[i] - nums[i - 1] > limit:
                parent_pos += 1
            store[parent_pos].append(nums[i])
            parent[nums[i]] = parent_pos
        
        for i in range(n):
            parent_pos = parent[nums_cp[i]]
            nums_cp[i] = store[parent_pos].popleft()

        return nums_cp